### ISSUE TITLE
Include user creation in ddl-resource.sql

### ISSUE THREAD

**Issue by rh-hemartin** [kubearchive/kubearchive#381]
In #378 there was a changed proposed to the installation flow. Instead of having the user provide the user and database for KubeArchive, the `ddl-resource.sql` should create the user and database instead.

**Comment by rh-hemartin** [comment]
I think a problem with this approach is that it assumes that the database user our user is provided has the permissions to create users and databases. I don't think that would be the case. However we can go with this approach and see if we need to change it later.

**Comment by maruiz93** [comment]
> a problem with this approach is that it assumes that the database user our user is provided has the permissions to create users and databases

I am not following here, so we are saying that we the KubeArchive, instead of:
1. Create a DB user for kubearchive
2. Create a database for kubearchive inside of the postgres
3. Apply the given DDL

Just do:
1. Apply the given DDL

The one that applies the DDL would be the same that creates a DB user and the DB, so I think it's a good idea to merge all the steps in just one.

I do not see the problem though (maybe I get lost with the different users of the sentence) can you rephrase it?

**Comment by maruiz93** [comment]
The only problem I see here is that the default password of the kubearchive user will be in the DDL too so the user will potentially need to change it in several places (secret and DDL)

**Comment by rh-hemartin** [comment]
> I do not see the problem though (maybe I get lost with the different users of the sentence) can you rephrase it?

The problem I see is that the database user that executes the SQL may not have permissions to create databases and other users.

> The one that applies the DDL would be the same that creates a DB user and the DB

I think this is incorrect: the one that needs to apply the DDL should be the KubeArchive Database User. One of the statements inside the DDL fail when applied by the admin user. See #384 .

> so we are saying that we the KubeArchive, instead of:

Just to be a little bit more specific, we are not executing the SQL. This is a SQL script meant to be run by the one who installs KubeArchive.

**Comment by maruiz93** [comment]
> I think this is incorrect: the one that needs to apply the DDL should be the KubeArchive Database User. One of the statements inside the DDL fail when applied by the admin user. See https://github.com/kubearchive/kubearchive/issues/384 .

I've reopened that issue and added a comment, PTAL.

> Just to be a little bit more specific, we are not executing the SQL. This is a SQL script meant to be run by the one who installs KubeArchive.

Yes we were on the same page here but I wasn't very clear so appreaciated the puntualization

**Comment by ggallen** [comment]
> I think this is incorrect: the one that needs to apply the DDL should be the KubeArchive Database User.

No, this is incorrect.  All the statements in the DDL file should be executed by a database administrator.  This user should have all the permissions.

**Comment by maruiz93** [comment]
As I have been doing some tests around this, I will pick this issue and propose a PR so we could discuss the changes over there

**Comment by ggallen** [comment]
I am going to propose a radical rewrite of this SQL.

First, I don't see any advantage to having our own schema. It just adds complexity. We can just stay in the public schema.  

We are trying to write SQL when there's no need for it, and it's causing problems. We should use `pg_dump` to generate this script.  That means the database is created in postgres as we want it, then we can dump it and have a perfectly generated script that will recreate it.  It's much less error prone.

I have taken our script, created the database, and then run pg_dump to generate the script. Here's the command:
```
pg_dump -U postgres -d kubearchive -C -c --if-exists -s
```
This created this script:
```
--
-- PostgreSQL database dump
--

-- Dumped from database version 16.1
-- Dumped by pg_dump version 16.1

SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET xmloption = content;
SET client_min_messages = warning;
SET row_security = off;

DROP DATABASE IF EXISTS kubearchive;

DROP ROLE IF EXISTS kubearchive;
CREATE USER kubearchive WITH ENCRYPTED PASSWORD 'P0stgr3sdbP@ssword';
--
-- Name: kubearchive; Type: DATABASE; Schema: -; Owner: kubearchive
--

CREATE DATABASE kubearchive WITH TEMPLATE = template0 ENCODING = 'UTF8' LOCALE_PROVIDER = libc LOCALE = 'en_US.UTF-8';


ALTER DATABASE kubearchive OWNER TO kubearchive;

\connect kubearchive

SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET xmloption = content;
SET client_min_messages = warning;
SET row_security = off;

--
-- Name: trigger_set_timestamp(); Type: FUNCTION; Schema: public; Owner: kubearchive
--

CREATE FUNCTION public.trigger_set_timestamp() RETURNS trigger
    LANGUAGE plpgsql
    AS $$
BEGIN
	  NEW.updated_at = NOW();
RETURN NEW;
END;
	$$;


ALTER FUNCTION public.trigger_set_timestamp() OWNER TO kubearchive;

SET default_tablespace = '';

SET default_table_access_method = heap;

--
-- Name: resource; Type: TABLE; Schema: public; Owner: kubearchive
--

CREATE TABLE public.resource (
    uuid uuid NOT NULL,
    api_version character varying NOT NULL,
    kind character varying NOT NULL,
    name character varying NOT NULL,
    namespace character varying NOT NULL,
    resource_version character varying,
    created_at timestamp without time zone DEFAULT now() NOT NULL,
    updated_at timestamp without time zone DEFAULT now() NOT NULL,
    cluster_deleted_ts timestamp without time zone,
    data jsonb NOT NULL
);


ALTER TABLE public.resource OWNER TO kubearchive;

--
-- Name: resource resource_pkey; Type: CONSTRAINT; Schema: public; Owner: kubearchive
--

ALTER TABLE ONLY public.resource
    ADD CONSTRAINT resource_pkey PRIMARY KEY (uuid);


--
-- Name: resource set_timestamp; Type: TRIGGER; Schema: public; Owner: kubearchive
--

CREATE TRIGGER set_timestamp BEFORE UPDATE ON public.resource FOR EACH ROW EXECUTE FUNCTION public.trigger_set_timestamp();


--
-- PostgreSQL database dump complete
--
```
The only think I added were the two command to DROP and CREATE the user after the database was dropped.

I think this is much cleaner and we should use this method going forward.

Another advantage is that you can just keep re-running this script over and over.  Since it drops everything first, everything gets recreated fresh.

**Comment by maruiz93** [comment]
> This created this script:

I personally prefer having less boilerplate, shorter DDL that we know what is doing, but I guess that is a matter of preference and maybe the rest prefer having this autogenerated DDL.

> Another advantage is that you can just keep re-running this script over and over. Since it drops everything first, everything gets recreated fresh.

I don't like this. Say that you already have a running version of kubearchive, and then, new version releases with an updated DDL file.

I don't think that we want to DROP the database, I think it's better to just create it if it isn't there.

I know that for that we should go through a different path regarding migrations, but for some changes (like adding a trigger) migration isn't needed and just rerunning the DDL could nicely work.

Also, I've tested it and the current DDL can be rerun again and again without failing.

**Comment by ggallen** [comment]
> I personally prefer having less boilerplate, shorter DDL that we know what is doing, but I guess that is a matter of preference and maybe the rest prefer having this autogenerated DDL.

Look at the time that has just been spent on errors in this file.  It's a waste, and totally unnecessary since we can have it created for us.

Can you really not read this file and understand what it is doing?  All the same commands are in there.

> Say that you already have a running version of kubearchive, and then, new version releases with an updated DDL file.

In that case I wouldn't apply this file, nor the existing one.  That's a migration which is different.

> I don't think that we want to DROP the database, I think it's better to just create it if it isn't there.

But what if the table schema has changed?  You won't get those changes.

> Also, I've tested it and the current DDL can be rerun again and again without failing.

I believe you can do this if you run it when you are connected to the right database.  But you have to be connected to it, or your table will be created in a database that you don't expect or want.

And I don't believe this is true. There is no check that the table exists, and if you try to create it when it already exists you will get an error.

**Comment by ggallen** [comment]
I just ran the existing DDL script twice:
```[gallen@gallen-thinkpadp1gen5 kubearchive] (upstream-main)$ psql -U postgres < database/ddl-resource.sql 
Password for user postgres: 
CREATE TABLE
CREATE FUNCTION
CREATE TRIGGER
ALTER TABLE
GRANT
ALTER FUNCTION
GRANT
GRANT
[gallen@gallen-thinkpadp1gen5 kubearchive] (upstream-main)$ psql -U postgres < database/ddl-resource.sql 
Password for user postgres: 
NOTICE:  relation "resource" already exists, skipping
CREATE TABLE
CREATE FUNCTION
ERROR:  trigger "set_timestamp" for relation "resource" already exists
ALTER TABLE
GRANT
ALTER FUNCTION
GRANT
GRANT
[gallen@gallen-thinkpadp1gen5 kubearchive] (upstream-main)$ 
```
I guess it succeeed, but it does say `ERROR`.

And to further test, I added a new column to the `resource` table and ran it again. The table creation really was skipped, and afterward my new column wasn't there.

**Comment by maruiz93** [comment]
> Look at the time that has just been spent on errors in this file. It's a waste, and totally unnecessary since we can have it created for us.

I don't think that there is that much time, I think that we have been spending time on thinking how the usability of the database would be, what should we do and don't, how should we provide it for testing and development... but specifically to that file, I don't think that much.

> Can you really not read this file and understand what it is doing? All the same commands are in there.

I understand it yes, but there are more variables in there that I think we don't need, it's not a big deal and that's why I say that if the rest think that they prefer the autogenerated that's ok.

> And to further test, I added a new column to the resource table and ran it again. The table creation really was skipped, and afterward my new column wasn't there.

Yes, that's why I said this  before:

>> I know that for that we should go through a different path regarding migrations, but for some changes (like adding a trigger) migration isn't needed and just rerunning the DDL could nicely work.

adding a column is a clear example of needing migrations, something that I think we should address separately, which you also mentioned in here:

>> In that case I wouldn't apply this file, nor the existing one. That's a migration which is different.

and I agree. I am refering to things like adding a new resource like a function or a trigger, or an index.

> I guess it succeeed, but it does say ERROR.

It succeeded yes, if the `ERROR` annoys you it's something we can address, but the script is doing what we expect


I think we can just defer this discussion to the Monday meeting

**Comment by mafh314** [comment]
> We are trying to write SQL when there's no need for it, and it's causing problems. We should use pg_dump to generate this script. That means the database is created in postgres as we want it, then we can dump it and have a perfectly generated script that will recreate it. It's much less error prone.

I agree with this. @ggallen the current ddl-resource.sql was created using the DLL tool from DBeaver, that tool set less default values than pg_dump. I'm ok, if we agreed, to use pg_dump.

**Comment by ggallen** [comment]
> I don't think that there is that much time

- https://github.com/kubearchive/kubearchive/issues/384
- https://github.com/kubearchive/kubearchive/issues/381
- https://github.com/kubearchive/kubearchive/issues/85
- https://github.com/kubearchive/kubearchive/pull/393
- https://github.com/kubearchive/kubearchive/pull/391

Four of these were created within the last 10 hours.  All deal, either directly or indirectly, with problems in the DDL file.

And it's not just one person, this has been a multi-person discussion. And then there will be reviews on the PRs, which again involve multiple people. Then comments and re-reviews.

So I agree in one sense - it's not a lot of time in the grand scheme of things. But it is time, and it's more than likely to reoccur in the future, multiple times.

**Comment by ggallen** [comment]
> the current ddl-resource.sql was created using the DLL tool from DBeaver

@mafh314 , thanks. I did not realize thatt, nd I doubt @maruiz93 did either.  But we've been modifying the file manually (see the PRs in the previous comment), which causes other problems. If we make changes manually, then we can never use the regenerated version again, unless we then go back an apply all the changes.  That's a nightmare.

I would opt for `pg_dump` just because it's from postgres and should always work.  As for the other `pg_dump` settings, I think it's good to have those in there.  It insures that the database created for KubeArchive is always the same.

And yes, I know I am proposing that this file be edited to add the `CREATE USER` command. That's just to include those two lines and have everything in one file.  It should always be the same, it it's the only manual change that should be made.  We could create a script that would insert them if that makes life easier.

**Comment by maruiz93** [comment]
@ggallen is correct I wasn't aware that the current DDL file were created through an automation, so I will close the current PR and leave the issue unassigned again.

@mafh314 or @ggallen feel free to pick it as you were already involved in these automation.
For the `CREATE USER` part, if you want to keep it reusable (the `IF NOT EXISTS` part), feel free to take a look at that part of my current PR.

> Four of these were created within the last 10 hours

You are mixing issues and their PRs in that list, that's a little unfair. 
There are only two PRs, and one of them is a bug of not adding the database in one of the resources that made the resource unreachable if not executed from it, a one-word fix... I think *this discussion* is the only one that has drain a little time from us.

The issue that doesn't have a PR on it, is not directly affecting the DDL, just how to apply it, so the automation doesn't affect it.

Anyway, there is no need to keep on this discussion, as I said before, I tend to avoid autogenerated code because there usually are more things than needed but it's not that much the difference and we were already using an automation so I'm ok with it.


Regarding doing DROP/CREATE instead of the equivalent of `CREATE IF NOT EXISTS`, I prefer the conservative approach, but it will probably be ok as for migrations we won't be reapplying the DDL.


Related to use the `public` schema, I haven't mentioned anything earlier but that's because I'm ok with that change, I actually don't have any argument against using public schema and in fact I don't really get the advantage of having different schemas in the database, I guess that I'm not used to work with them.

**Comment by maruiz93** [comment]
This issue was addressed in https://github.com/kubearchive/kubearchive/pull/417
